File: python-backend/app/services/health_service.py
# FILE: app/services/health_service.py
# JOB: Actually test the Python backend's connections.

import time
import redis
from app.config.supabase_client import supabase

# How long we wait before calling a check failed.
TIMEOUT_SECONDS = 3


# ---- The simple check. Only proves the server is answering. ----
def get_health_status():
    return {
        "status": "ok",
        "service": "python-ai-backend",
    }


# ---- Check 1: the database ----
def check_database():
    started = time.time()
    try:
        # Ask for one row. This proves the connection works.
        supabase.table("profiles").select("id").limit(1).execute()
        return {
            "name": "database",
            "status": "up",
            "required": True,
            "ms": int((time.time() - started) * 1000),
        }
    except Exception as error:
        return {
            "name": "database",
            "status": "down",
            "required": True,
            "ms": int((time.time() - started) * 1000),
            "error": str(error),
        }


# ---- Check 2: Redis (used for the AI cache) ----
def check_redis():
    started = time.time()
    try:
        # Make our own connection with a short timeout,
        # so a frozen Redis cannot hang this check.
        client = redis.Redis(
            host="redis",
            port=6379,
            socket_connect_timeout=TIMEOUT_SECONDS,
            socket_timeout=TIMEOUT_SECONDS,
        )
        if not client.ping():
            raise Exception("Redis did not reply")
        return {
            "name": "redis_cache",
            "status": "up",
            "required": True,
            "ms": int((time.time() - started) * 1000),
        }
    except Exception as error:
        return {
            "name": "redis_cache",
            "status": "down",
            "required": True,
            "ms": int((time.time() - started) * 1000),
            "error": str(error),
        }


# ---- Check 3: is the AI model server address set? ----
# We do NOT call Colab here. Waking it up on every health check
# would be slow and wasteful. We only check the setting exists.
def check_ai_config():
    import os

    colab_url = os.environ.get("COLAB_AI_URL", "")
    if colab_url:
        return {
            "name": "ai_models",
            "status": "up",
            "required": False,
            "note": "Address is set. Not called, to avoid waking it.",
        }
    return {
        "name": "ai_models",
        "status": "down",
        "required": False,
        "error": "COLAB_AI_URL is not set. The AI senses will not work.",
    }


# ---- Check 4: is the OpenAI key set? ----
def check_openai_config():
    import os

    key = os.environ.get("OPENAI_API_KEY", "")
    if key:
        return {
            "name": "openai",
            "status": "up",
            "required": False,
            "note": "Key is set. Not called, to avoid cost.",
        }
    return {
        "name": "openai",
        "status": "down",
        "required": False,
        "error": "OPENAI_API_KEY is not set. The brain will use rules only.",
    }


# ---- Work out the overall state ----
# Pure logic, easy to test.
def overall_status(checks):
    # Any required piece down means the whole thing is down.
    for check in checks:
        if check.get("required") and check.get("status") == "down":
            return "down"

    # An optional piece down means degraded, not down.
    for check in checks:
        if not check.get("required") and check.get("status") == "down":
            return "degraded"

    return "up"


# ---- Run every check and build the report ----
def check_all():
    started = time.time()

    checks = [
        check_database(),
        check_redis(),
        check_ai_config(),
        check_openai_config(),
    ]

    return {
        "status": overall_status(checks),
        "service": "python-ai-backend",
        "total_ms": int((time.time() - started) * 1000),
        "checks": checks,
    }

File: node-backend/src/services/healthService.js
// FILE: src/services/healthService.js
// JOB: Actually test every piece of the system, and report the truth.

const axios = require("axios");
const supabase = require("../config/supabase");
const redisClient = require("../config/redis");
const mqttClient = require("../config/mqtt");

// How long we wait before calling a check failed.
// A health check must never hang, or it becomes the problem itself.
const TIMEOUT_MS = 3000;

// ---- A helper that gives up after a set time ----
// It races the real task against a timer. Whichever finishes first wins.
function withTimeout(promise, label) {
   const timer = new Promise(function (resolve, reject) {
      setTimeout(function () {
         reject(new Error(label + " timed out after " + TIMEOUT_MS + "ms"));
      }, TIMEOUT_MS);
   });

   return Promise.race([promise, timer]);
}

// ---- Check 1: the database ----
async function checkDatabase() {
   const startedAt = Date.now();

   try {
      // Ask for one row. This proves the connection really works.
      // "head: true" means do not send the row back, only check it worked.
      const query = supabase
         .from("profiles")
         .select("id", { count: "exact", head: true });

      await withTimeout(query, "database");

      return {
         name: "database",
         status: "up",
         required: true,
         ms: Date.now() - startedAt,
      };
   } catch (error) {
      return {
         name: "database",
         status: "down",
         required: true,
         ms: Date.now() - startedAt,
         error: error.message,
      };
   }
}

// ---- Check 2: Redis ----
async function checkRedis() {
   const startedAt = Date.now();

   try {
      // "ping" is Redis's own way of saying "are you there".
      // A healthy Redis replies with the word PONG.
      const reply = await withTimeout(redisClient.ping(), "redis");

      if (reply !== "PONG") {
         throw new Error("Redis gave an unexpected reply");
      }

      return {
         name: "redis",
         status: "up",
         required: true,
         ms: Date.now() - startedAt,
      };
   } catch (error) {
      return {
         name: "redis",
         status: "down",
         required: true,
         ms: Date.now() - startedAt,
         error: error.message,
      };
   }
}

// ---- Check 3: the MQTT broker ----
function checkMqtt() {
   const startedAt = Date.now();

   // The mqtt library keeps a "connected" flag for us.
   // No network call is needed, so this is instant.
   if (mqttClient.connected) {
      return {
         name: "mqtt_broker",
         status: "up",
         required: true,
         ms: Date.now() - startedAt,
      };
   }

   return {
      name: "mqtt_broker",
      status: "down",
      required: true,
      ms: Date.now() - startedAt,
      error: "Not connected to the broker",
   };
}

// ---- Check 4: the Python backend ----
async function checkPythonBackend() {
   const startedAt = Date.now();

   try {
      // Ask Python's own deep health endpoint.
      const response = await axios.get("http://python-backend:8000/health/deep", {
         timeout: TIMEOUT_MS,
      });

      // Python tells us its own overall state. We pass it through.
      const pythonStatus = response.data.status;

      return {
         name: "python_backend",
         status: pythonStatus === "up" ? "up" : "degraded",
         required: true,
         ms: Date.now() - startedAt,
         detail: response.data,
      };
   } catch (error) {
      return {
         name: "python_backend",
         status: "down",
         required: true,
         ms: Date.now() - startedAt,
         error: error.message,
      };
   }
}

// ---- Work out the overall state from all the checks ----
// Pure logic, so it is easy to test.
function overallStatus(checks) {
   // If any REQUIRED piece is down, the whole system is down.
   for (let i = 0; i < checks.length; i++) {
      if (checks[i].required && checks[i].status === "down") {
         return "down";
      }
   }

   // If anything is degraded, or an optional piece is down,
   // the system works but not fully.
   for (let i = 0; i < checks.length; i++) {
      if (checks[i].status === "degraded") {
         return "degraded";
      }
      if (!checks[i].required && checks[i].status === "down") {
         return "degraded";
      }
   }

   return "up";
}

// ---- Run every check and build the report ----
async function checkAll() {
   const startedAt = Date.now();

   // Run the slow checks AT THE SAME TIME, not one after another.
   // Promise.all waits for all of them together.
   const results = await Promise.all([
      checkDatabase(),
      checkRedis(),
      checkPythonBackend(),
   ]);

   // The MQTT check is instant, so we just add it.
   results.push(checkMqtt());

   return {
      status: overallStatus(results),
      service: "node-backend",
      checked_at: new Date().toISOString(),
      total_ms: Date.now() - startedAt,
      checks: results,
   };
}

module.exports = {
   checkAll: checkAll,
   overallStatus: overallStatus,
};
